lib/sneakers_handlers/exponential_backoff_handler.rb

```diff
@@ -0,0 +1,121 @@
+module SneakersHandlers
+  class ExponentialBackoffHandler
+    attr_reader :queue, :channel, :options, :delays
+
+    def initialize(channel, queue, options)
+      @queue = queue
+      @channel = channel
+      @options = options
+      @delays = options[:delays] || [1.second, 10.seconds, 1.minute, 10.minutes]
+
+      Array(@options[:routing_key]).each do |key|
+        queue.bind(primary_exchange, routing_key: key + ".*")
+      end
+    end
+
+    def acknowledge(delivery_info, _, _)
+      @channel.acknowledge(delivery_info.delivery_tag, false)
+    end
+
+    def reject(delivery_info, properties, message)
+      retry_message(delivery_info, properties, message, :reject)
+    end
+
+    def error(delivery_info, properties, message, err)
+      retry_message(delivery_info, properties, message, err)
+    end
+
+    def timeout(delivery_info, properties, message)
+      retry_message(delivery_info, properties, message, :timeout)
+    end
+
+    def noop(delivery_info, properties, message); end
+
+    private
+
+    def retry_message(delivery_info, properties, message, reason)
+      attempt_number = death_count(properties[:headers])
+
+      routing_key_segments = delivery_info[:routing_key].split(".")
+      routing_key_segments.pop if Integer(routing_key_segments.last) rescue nil
+
+      if attempt_number < @delays.length
+        delay = @delays[attempt_number]
+
+        log("msg=retrying, delay=#{delay}, count=#{attempt_number}, properties=#{properties}")
+
+        routing_key_segments << delay
+        routing_key = routing_key_segments.join(".")
+
+        retry_queue = create_retry_queue!(delay)
+        retry_queue.bind(retry_exchange, routing_key: routing_key)
+
+        retry_exchange.publish(message, routing_key: routing_key, headers: properties[:headers])
+      else
+        log("msg=erroring, count=#{attempt_number}, properties=#{properties}")
+
+        error_exchange.publish(message, routing_key: delivery_info[:routing_key])
```

Comment:
Hm, I think we can deliver messages to the wrong queue using this routing key. In cases where we have two queues bound to the same routing key, when the message fails in one queue, it'd be sent to both dead letter queues, right? 
Can we use `x-dead-letter-routing-key` instead?